mlflow/utils/gorilla.py

```diff
@@ -0,0 +1,898 @@
+
+    Parameters
+    ----------
+    patch : gorilla.Patch
+        Patch.
+
+    Raises
+    ------
+    RuntimeError
+        Overwriting an existing attribute is not allowed when the setting
+        :attr:`Settings.allow_hit` is set to ``True``.
+
+    Note
+    ----
+    If both the attributes :attr:`Settings.allow_hit` and
+    :attr:`Settings.store_hit` are ``True`` but that the target attribute seems
+    to have already been stored, then it won't be stored again to avoid losing
+    the original attribute that was stored the first time around.
+    """
+    settings = Settings() if patch.settings is None else patch.settings
+
+    # When a hit occurs due to an attribute at the destination already existing
+    # with the patch's name, the existing attribute is referred to as 'target'.
+    try:
+        target = get_attribute(patch.destination, patch.name)
+    except AttributeError:
+        pass
+    else:
+        if not settings.allow_hit:
+            raise RuntimeError(
+                "An attribute named '%s' already exists at the destination "
+                "'%s'. Set a different name through the patch object to avoid "
+                "a name clash or set the setting 'allow_hit' to True to "
+                "overwrite the attribute. In the latter case, it is "
+                "recommended to also set the 'store_hit' setting to True in "
+                "order to store the original attribute under a different "
+                "name so it can still be accessed." % (patch.name, patch.destination.__name__)
+            )
+
+        if settings.store_hit:
+            original_name = _ORIGINAL_NAME % (patch.name,)
+            # For certain MLflow Models use cases, such as scikit-learn autologging, we patch
+            # a method on a parent class
+            # (e.g., `sklearn.feature_extraction.text.CountVectorizer.fit_transform()`) and
+            # subsequently patch a corresponding overridden method on one of its children
+            # (e.g., `feature_extraction.text.TfidfVectorizer.fit_transform()`)
+            # to provide a different implementation. In these cases, we wish to swap out the
+            # "original function" attribute stored on the child in order to refer to the child's
+            # overriden method (e.g., `feature_extraction.text.TfidfVectorizer.fit_transform()`),
+            # rather than the parent method
+            # (e.g., `sklearn.feature_extraction.text.CountVectorizer.fit_transform()`)
+            prev_patch = getattr(patch.destination, _ACTIVE_PATCH, None)
+            if not hasattr(patch.destination, original_name) or (
+                prev_patch
+                and prev_patch.destination != patch.destination
+                and issubclass(patch.destination, prev_patch.destination)
+            ):
+                setattr(patch.destination, original_name, target)
+
+    setattr(patch.destination, patch.name, patch.obj)
+    setattr(patch.destination, _ACTIVE_PATCH, patch)
```

Comment:
@harupy These are the critical portions of the change to address problem (1) from https://github.com/mlflow/mlflow/issues/3574#issuecomment-714994583.

We now retain a reference to the previous patch that was applied to a class or one of its super classes, and, if  the patch was applied to a super class, we overwrite it with the patch that's being applied to this class directly.